nxc/protocols/ssh.py

```diff
@@ -63,46 +70,137 @@ def create_conn_obj(self):
+            # check if user can execute mkfifo
+            if "Create named pipes" in stdout:
+                self.logger.info("Command: 'mkfifo' available")
+                pipe_stdin = f'/tmp/systemd-{uuid.uuid4()}'
+                pipe_stdout = f'/tmp/systemd-{uuid.uuid4()}'
+                shadow_Backup = f'/tmp/{uuid.uuid4()}'
+                stdin, stdout, stderr = self.conn.exec_command(f"mkfifo {pipe_stdin}; tail -f {pipe_stdin} | /bin/sh 2>&1 > {pipe_stdout} >/dev/null 2>&1 &")
+                # 'script -qc /bin/sh /dev/null' means "upgrade" the shell, like reverse shell from netcat
+                stdin, stdout, stderr = self.conn.exec_command(f"echo 'script -qc /bin/sh /dev/null' > {pipe_stdin}")
+                stdin, stdout, stderr = self.conn.exec_command(f"echo 'sudo -s' > {pipe_stdin} && echo '{self.password}' > {pipe_stdin}")
+                # Sometime the pipe will hanging(only happen with paramiko)
+                # Can't get "whoami" or "id" result in pipe_stdout, maybe something wrong using pipe with paramiko
+                # But one thing I can confirm, is the command was executed even can't get result from pipe_stdout
+                tries = 1
+                self.logger.info(f"Copy /etc/shadow to {shadow_Backup} if pass the sudo auth")
+                while True:
+                    self.logger.info(f"Checking {shadow_Backup} if it existed")
+                    stdin, stdout, stderr = self.conn.exec_command(f'ls {shadow_Backup}')
+                    if tries >= self.args.get_output_tries:
+                        self.logger.info(f'{shadow_Backup} not existed, maybe the pipe has been hanged over, please increase the number of tries with the option "--get-output-tries" or change other method with "--sudo-check-method". If it\'s still failing maybe sudo shell is not working with current user')
+                        break
+
+                    if stderr.read().decode('utf-8'):
+                        time.sleep(2)
+                        stdin, stdout, stderr = self.conn.exec_command(f"echo 'cp /etc/shadow {shadow_Backup} && chmod 777 {shadow_Backup}' > {pipe_stdin}")
+                        tries += 1
+                    else:
+                        self.logger.info(f"{shadow_Backup} existed")
+                        self.admin_privs = True
+                        break
+                self.logger.info(f"Remove up temporary files")
+                stdin, stdout, stderr = self.conn.exec_command(f"rm -rf {shadow_Backup} {pipe_stdin} {pipe_stdout}")
+            else:
+                self.logger.error("Command: 'mkfifo' unavailable, running command with 'sudo' failed")
+                return
 
     def plaintext_login(self, username, password, private_key=None):
+        self.username = username
+        self.password = password
+        pkey = ""
+        stdout = None
+        stderr = None
+        cred_id = self.db.add_credential("plaintext", username, password)
```

Comment:
This should be placed after the login with username&password to ensure only valid credentials are added to the db

Reply:
Fixed